core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,382 @@
+                          expressionEvaluationContext: ExpressionEvaluationContext,
+                          gatherInputsResult: MethodConfigResolver.GatherInputsResult
+  )(implicit executionContext: ExecutionContext): Future[LazyList[SubmissionValidationEntityInputs]] = {
+    // First, verify that necessary entity information is present and consistent
+    val entityType = expressionEvaluationContext.entityType.getOrElse(
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing entityType"))
+    )
+    val entityName = expressionEvaluationContext.entityName.getOrElse(
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing entityName"))
+    )
+
+    val rootEntityTypeOpt = expressionEvaluationContext.rootEntityType
+    if (rootEntityTypeOpt.isEmpty) {
+      Future.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing rootEntityType")))
+    } else if (
+      expressionEvaluationContext.expression.isEmpty &&
+      entityType != rootEntityTypeOpt.get
+    ) {
+      val whatYouGaveUs =
+        if (expressionEvaluationContext.entityType.isDefined)
+          s"an entity of type ${expressionEvaluationContext.entityType.get}"
+        else "no entity"
+      Future.failed(
+        new RawlsExceptionWithErrorReport(
+          errorReport = ErrorReport(
+            StatusCodes.BadRequest,
+            s"Method configuration expects an entity of type ${rootEntityTypeOpt.get}, but you gave us $whatYouGaveUs."
+          )
+        )
+      )
+    } else {
+
+      // Next, parse both the entity expression (if it exists) and the input expressions
+      // To determine which entities to fetch from the database
+      val rootEntityType = rootEntityTypeOpt.get
+
+      val entityLookups: Seq[ExpressionLookup] = expressionEvaluationContext.expression match {
+        case None             => Seq.empty
+        case Some(expression) => parseLookups(expression)
+      }
+
+      // Parse all input expressions and collect their lookups and parsed trees
+      val inputExpressionData = gatherInputsResult.processableInputs.toSeq.map { input =>
+        val terraExpressionParser = AntlrTerraExpressionParser.getParser(input.expression)
+        val visitor = new CompactEvaluateVisitor()
+        val parsedTree = terraExpressionParser.root()
+        val inputLookups: Seq[ExpressionLookup] = visitor.visit(parsedTree)
+        (input, parsedTree, inputLookups)
+      }
+
+      // Gather all ExpressionLookups from all inputs
+      val allLookups = inputExpressionData.flatMap(_._3)
+
+      // Build query plans for all lookups combined
+      val queryPlans = if (entityType != rootEntityType && entityLookups.nonEmpty) {
+        // Use the entityLookups to determine the starting point for the relationChains in the query plans
+        val entityRelationChain = entityLookups.flatMap(_.relations.map(_.attributeName().getText)).toList
```

Comment:
I am not sure what `.attributeName().getText` returns here in the case of an attribute with non-default namespace like `import:timestamp`.  You might have to do something ugly verbose like
```scala
_.relations.map(relCtx =>
              AttributeName.toDelimitedName(
                AttributeName(relCtx.attributeName().namespace().getText, relCtx.attributeName().name().getText)
              )
            )
```